ffmpeg-plugin/mcm_audio_tx.c

```diff
@@ -114,29 +117,50 @@ static int mcm_audio_write_header(AVFormatContext* avctx)
 static int mcm_audio_write_packet(AVFormatContext* avctx, AVPacket* pkt)
 {
     McmAudioMuxerContext *s = avctx->priv_data;
+    size_t frame_size = s->tx_handle->frame_size;
     const uint8_t *data = pkt->data;
-    mcm_buffer *buf = NULL;
     int size = pkt->size;
     int len, err = 0;
 
     while (size > 0) {
-        buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
-        if (buf == NULL) {
-            av_log(avctx, AV_LOG_ERROR, "Dequeue buffer error %d\n", err);
-            return AVERROR(EIO);
+        char *dest;
+
+        if (!s->unsent_mcm_buf) {
+            s->unsent_mcm_buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
+            if (s->unsent_mcm_buf == NULL) {
+                av_log(avctx, AV_LOG_ERROR, "Initial dequeue buffer error %d\n", err);
+                return AVERROR(EIO);
+            }
         }
 
-        len = FFMIN(buf->len, size);
-        memcpy(buf->data, data, len);
+        len = FFMIN(frame_size, size + s->unsent_len);
+
+        if (len < frame_size) {
+            memcpy((char *)s->unsent_mcm_buf->data + s->unsent_len, data, size);
+            s->unsent_len += size;
+            break;
+        }
+
+        dest = s->unsent_mcm_buf->data;
+
+        if (s->unsent_len) {
+            memcpy(dest, s->unsent_mcm_buf->data, s->unsent_len);
```

Comment:
This is literally `memcpy(dest, dest, s->unsent_len);` so undefined behavior of memcpy for overlapping memory slices.

Reply:
OMG. Great catch Milosz! Let me rework that.

Reply:
This memcpy has to be removed. It left after refactoring but didn't affect the algorithm. That's why I missed it. I'm going to fix it in a separate commit.